valhalla/baldr/directededge.h

```diff
@@ -1121,7 +1127,8 @@ class DirectedEdge {
   uint64_t seasonal_ : 1;       // Seasonal access (ex. no access in winter)
   uint64_t deadend_ : 1;        // Leads to a dead-end (no other driveable roads) TODO
   uint64_t bss_connection_ : 1; // Does this lead to(come out from) a bike share station?
-  uint64_t spare4_ : 9;
+  uint64_t max_speed_ : 8;      // speed_ can be maxspeed:advisory or maxspeed:practical
+  uint64_t spare4_ : 1;
```

Comment:
oh man... this is really rough. i'm having a hard time making a case for this just because of advisory or practical. i'd almost rather just stop parsing advisory or practical so we can use `speed_` instead.

@dnesbitt61 or @gknisely your thoughts?

Reply:
Haha no worries, I figured so! An alternative could be to change the `SpeedType` enum to `kTagDefault`, `kTagPractical`, `kTagMax`, `kClassified`:

https://github.com/valhalla/valhalla/blob/nn_legal_limit/valhalla/baldr/graphconstants.h#L365-L369

Plus the code to parse them from the PBF.. And reverse however `speed` was enhanced after parsing. That way we could at least detect if the tagged speed is from a `maxspeed` tag and not one of the others and not write too low speeds to `max_speed_`..

Or chop default & practical ;) Or a no-merge.. up to you.

Reply:
speed_type in DirectedEdge is only 1 bit so this doesn't seem practical.

Reply:
i had this thought too. actually i think we **can** do it by taking a single free bit, shifting it one and oring it with the existing bit. effectively making this speed type field 2 bits wide. it would almost be backwards compatible in that old data would work with the new code. but new data with old code would mean that anything that were advisory or practical speed types would look like categorized speed types. i feel like this is a livable compromise since since they are sparsely populated in the data.

Reply:
> i feel like this is a livable compromise

totally agreed!

Reply:
Shortcuts would also need to know their speed type then and so far that's not recorded.

I guess a shortcut should only get a speed limit assigned if all participating edges had one, otherwise default to 0 as now. If all had one, the speed limit should be the minimum of all edges' maxspeed. Correct me if I'm wrong though.

Reply:
Though I'm not sure if the edges making up a shortcut are even available to the routing algos? In the case where a shortcut speed limit would exceed a `legal_speed` one might still legally travel _some_ root edges of the shortcut since they could have higher speed limits. Should be pretty rare and can probably be disregarded, but it'd also help my general understanding..

Reply:
Ah forget it.. just found the spot that skips building shortcuts if their speeds are different: 

https://github.com/valhalla/valhalla/blob/d3c4eece264ab4ca846fb373cf75ef14dcd1e1d2/src/mjolnir/shortcutbuilder.cc#L255-L258

Reply:
would driving routes need to be impacted by the max speed logic? If not then it seems unnecessary to check speed limits when creating shortcuts. But then, truck routing (for instance) uses shortcuts but doesn't have a similar check for truck speed here - seems like it should. Problem is the more attributes you use in costing, the less effective shortcuts become.

Reply:
Ugh sorry, I'm getting confused here anyways.. With the new approach (2 bit SpeedType) `speed` is what will hold the maxspeed, not `speed_limit` (not accessible to DirectedEdge). So shortcuts are fine actually..

Tbh, I don't really understand why different speeds are a reason to not create a shortcut @dnesbitt61 .. Couldn't the `top_speed' costing work with average speeds on the accumulated shortcut? The same for setting a `speed_` on the final shortcut DirectedEdge, no? Probably I'm missing a crucial step here, but I can't see it.. Supporting a legal speed limit for motorized costings would need some logic like that though, but it could be limited to `maxspeed` tagged ones.

Reply:
I take back my comment on truck routing and truck speed - it doesn't look like truck speed is used when doing truck routing which seems wrong.
Wouldn't any costing that uses some penalty or avoidance when speed is above some threshold need to make sure that shortcuts don't filter out such speeds? A connected set of edges that are 45MPH, then 55MPH, then back to 45MPH would need to keep the 55MPH section separate if any costing penalizes edges above 45MPH. Alternatively, any costing which penalizes or avoids edges by speed might need to avoid using shortcut edges.

Reply:
> it doesn't look like truck speed is used when doing truck routing which seems wrong

actually it's used when `GetSpeed()` is called in the costing, there's a (fairly new I think) `is_truck` bool arg. 

> A connected set of edges that are 45MPH, then 55MPH, then back to 45MPH would need to keep the 55MPH section separate if any costing penalizes edges above 45MPH

yeah, of course..

Reply:
been doing only bicycle routing recently and haven't followed recent updates to speeds and their use (or shortcuts for that matter). Sorry for wading into this discussion without full knowledge!

Reply:
All good, you cleared up some misconceptions on my side:)